File: 30DaysOfCode/day18_queuesAndStacks/queuesAndStacks.java
import java.io.*;
import java.util.*;

public class Solution {
    private static Stack<Character> s = new Stack<Character>();
    private static Queue<Character> q = new LinkedList<Character>();
    
    public static void pushCharacter(char c) {
        s.push(new Character(c));
    }
    
    public static void enqueueCharacter(char c) {
        q.add(new Character(c));
    }
    
    public static char popCharacter() {
        return (char)s.pop();
    }
    
    public static char dequeueCharacter() {
        return (char)q.remove();
    }
    
    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        String input = scan.nextLine();
        scan.close();

        char[] s = input.toCharArray();
        Solution p = new Solution();

        for (char c : s) {
            p.pushCharacter(c);
            p.enqueueCharacter(c);
        }

        boolean isPalindrome = true;
        for (int i = 0; i < s.length/2; i++) {
            if (p.popCharacter() != p.dequeueCharacter()) {
                isPalindrome = false;                
                break;
            }
        }

        System.out.println( "The word, " + input + ", is " + ( (!isPalindrome) ? "not a palindrome." : "a palindrome." ) );
    }
}

File: 30DaysOfCode/day18_queuesAndStacks/queuesAndStacks.rb
class Solution
    attr_accessor :s, :q
    
  	def initialize()
        @s = []
        @q = []
    end
    
    def push_character(c)
        @s.push c
    end
    
    def enqueue_character(c)
        @q.push c
    end
    
    def pop_character()
        return @s.pop
    end
    
    def dequeue_character()
        return @q.shift
    end
end

solution = Solution.new
input = gets

input.split('').each do |c|
    solution.push_character c
    solution.enqueue_character c
end

is_palindrome = true

(input.length / 2).times do
    if solution.pop_character != solution.dequeue_character
        is_palindrome = false
        break
    end
end

if is_palindrome
    puts "The word, #{input}, is a palindrome."
else
    puts "The word, #{input}, is not a palindrome."
end
